src/main/kotlin/me/zeroeightsix/kami/util/MessageDetectionHelper.kt

```diff
@@ -0,0 +1,61 @@
+package me.zeroeightsix.kami.util
+
+import java.util.regex.Pattern
+
+object MessageDetectionHelper {
+    fun String.detectAndRemove(regex: Regexes): String? {
+        if (!this.detect(regex)) return null
+
+        val removed = regex.regex.replace(this, "")
+        return if (removed.isEmpty()) null else {
+            removed
+        }
+    }
+
+    fun String.detect(vararg regexes: Regexes): Boolean {
+        regexes.forEach {
+            if (it.regex.containsMatchIn(this)) return true
+        }
+        return false
+    }
+
+    fun String.detect(setting: Boolean, vararg regex: Regexes) = setting && this.detect(*regex)
+
+    fun isDirectReceived(direct: Boolean, message: String) =
+        message.detect(direct, Regexes.DIRECT, Regexes.DIRECT_ALT_1, Regexes.DIRECT_ALT_2)
+
+    fun isDirect(direct: Boolean, message: String) =
+        message.detect(direct, Regexes.DIRECT_SENT) || isDirectReceived(direct, message)
+
+    fun getDirectUsername(message: String): String? {
+        if (!isDirect(true, message)) return null
+
+        /* side note: this won't work if some glitched account has spaces in their username, but in all honesty, like 3 people globally have those */
+        val split = message.split(" ")
+        var username = split[0]
+
+        if (message.detect(Regexes.DIRECT_ALT_1) && username.length > 1) {
+            username = username.substring(1) // remove preceding [
+        }
+
+        if (message.detect(Regexes.DIRECT_ALT_2)) {
+            username = split[1].dropLast(1) // remove trailing :
+        }
+
+        return username
```

Comment:
This is terrible. Please use named regex groups: 
```regex
<(?<username>[[:ascii:]]+)> (?<message>.*)
```
for messages
```
<086> Hi, what's up?
<l1ving> I really like regex.
```
Will provide access to named groups:
```
username    message
086         Hi, what's up?
l1ving      I really like regex.
```

If the regex 'may' match unecessary characters, modify your regex not to do that instead.